Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp

```diff
@@ -0,0 +1,268 @@
+
+std::shared_ptr<ParameterLink<bool>> LexicaseOptimizer::recordOptimizeValuesPL = 
+Parameters::register_parameter("OPTIMIZER_LEXICASE-recordOptimizeValues", true,
+	"record optimize values to data files using optimizeFormulaNames");
+
+
+LexicaseOptimizer::LexicaseOptimizer(std::shared_ptr<ParametersTable> PT_)
+    : AbstractOptimizer(PT_) {
+
+	std::vector<std::string> optimizeFormulasStrings;
+	convertCSVListToVector(optimizeFormulasPL->get(PT), optimizeFormulasStrings);
+
+	for (auto s : optimizeFormulasStrings) {
+		optimizeFormulasMTs.push_back(stringToMTree(s));
+	}
+
+	// get names to use with scores
+	if (optimizeFormulaNamesPL->get(PT) == "default") {
+		// user has not defined names, auto generate names
+		for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
+			scoreNames.push_back("optimizeValue_" + std::to_string(fIndex));
+		}
+	}
+	else {
+		// user has defined names, use those
+		convertCSVListToVector(optimizeFormulaNamesPL->get(PT), scoreNames);
+	}
+
+
+	epsilon = epsilonPL->get(PT);
+	epsilonByRange = epsilonByRangePL->get(PT);
+	tournamentSize = tournamentSizePL->get(PT);
+	nextPopSizeFormula = stringToMTree(nextPopSizePL->get(PT));
+	numberParents = numberParentsPL->get(PT);
+	recordOptimizeValues = recordOptimizeValuesPL->get(PT);
+	
+	// leave this undefined so that max.csv is not generated
+	//optimizeFormula = optimizeValueMT;
+
+	popFileColumns.clear();
+	if (recordOptimizeValues){
+		for (auto name : scoreNames) {
+			popFileColumns.push_back(name);
+		}
+	}
+}
+
+
+int LexicaseOptimizer::lexiSelect(const std::vector<int> &tournamentIndexList) {
+	if (!scoresHaveDelta) { // if all scores are the same! pick random
+		return Random::getIndex(tournamentIndexList.size());
+	}
+
+	// generate a vector with formulasSize and fill with values from 0 to formulasSize in a random order.
+	std::vector<int> formulasOrder;
+	std::vector<int> sourceValues; // temp vector used to generate formulasOrder
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		sourceValues.push_back(i);
+	}
+	for (size_t i = 0; i < optimizeFormulasMTs.size(); i++) {
+		int pick = Random::getIndex(sourceValues.size());
+		formulasOrder.push_back(sourceValues[pick]);
+		sourceValues[pick] = sourceValues.back();
+		sourceValues.pop_back();
+	}
+	// now we have a random order in formulasOrder
+
+	// keepers is the current list of indexes into population for orgs which have passed all tests.
+	std::vector<int> keepers  = tournamentIndexList;
+
+	while (keepers.size() > 1 && formulasOrder.size() > 0) {
+		// while there are still atleast one keeper and there are still formulas
+		int formulaIndex = formulasOrder.back();
+		formulasOrder.pop_back();
+
+		double scoreCutoff;
+
+		if (epsilonByRange) {
+			// get the range of scores
+			double maxScoreHere = scores[keepers[0]][formulaIndex];
+			double minScoreHere = maxScoreHere;
+			for (size_t i = 0; i < keepers.size(); i++) {
```

Comment:
This should be std::min_element and std::max_element

Reply:
I think this would require the same type of restructuring as remove_if